Sources/swift-format/Frontend/Frontend.swift

```diff
@@ -124,28 +124,41 @@ class Frontend {
   }
 
   /// Processes source content from a list of files and/or directories provided as paths.
-  private func processPaths(_ paths: [String]) {
+  private func processPaths(_ paths: [String], parallel: Bool) {
     precondition(
       !paths.isEmpty,
       "processPaths(_:) should only be called when paths is non-empty.")
 
-    for path in FileIterator(paths: paths) {
-      guard let sourceFile = FileHandle(forReadingAtPath: path) else {
-        diagnosticEngine.diagnose(Diagnostic.Message(.error, "Unable to open \(path)"))
-        continue
+    if parallel {
+      let allFilePaths = Array(FileIterator(paths: paths))
```

Comment:
_Rambling:_ I still have some concerns about the upfront flattening of the `FileIterator` being slow if the formatter is run in parallel on something like a networked file system or in recursive mode on the root of a large monorepo (who would do such a thing), but I should probably test my concerns against real data before going to the trouble of adding more work.

If we need to change this in the future, I was concerned about having to do our own thread pooling, but maybe there's an easy alternative here: instead of flattening the whole iterator at once, just pull off a both sufficiently large and sufficiently small number of files at a time (say, 200, or a nice round number like 256) and loop over the batches while processing each batch with `concurrentPerform`. That would probably still get us good parallelism in the batches while avoiding eagerly flattening the whole array.

You don't need to make this change now (I might be worried about the performance for nothing), but I just wanted to write it out here so I don't forget it later if we do need it.

Reply:
+1 good thoughts, I would definitely be interested to hear if you see issues with this given the large number of files.